llvm/test/tools/llvm-symbolizer/skip-line-zero.s

```diff
@@ -0,0 +1,252 @@
+	.long	0x16df-0x16c0                   # DW_AT_high_pc (.Lfunc_end1-.Lfunc_begin1)
+	.byte	1                               # DW_AT_frame_base
+	.byte	86
+	.long	.Linfo_string4                  # DW_AT_name
+	.byte	1                               # DW_AT_decl_file
+	.byte	2                               # DW_AT_decl_line
+                                        # DW_AT_prototyped
+	.long	88                              # DW_AT_type
+                                        # DW_AT_external
+	.byte	4                               # Abbrev [4] 0x58:0x7 DW_TAG_base_type
+	.long	.Linfo_string3                  # DW_AT_name
+	.byte	5                               # DW_AT_encoding
+	.byte	4                               # DW_AT_byte_size
+	.byte	0                               # End Of Children Mark
+.Ldebug_info_end0:
+	.section	.debug_ranges,"",@progbits
+.Ldebug_ranges0:
+	.quad	0x1710                          #.Lfunc_begin0
+	.quad	0x171b                          #.Lfunc_end0
+	.quad	0x16c0                          #.Lfunc_begin1
+	.quad	0x16df                          #.Lfunc_end1
+	.quad	0
+	.quad	0
+	.section	.debug_str,"MS",@progbits,1
+.Linfo_string0:
+	.asciz	""                              # string offset=0
+.Linfo_string1:
+	.asciz	"main.c"                        # string offset=113
+.Linfo_string2:
+	.asciz	"foo"                           # string offset=120
+.Linfo_string3:
+	.asciz	"int"                           # string offset=124
+.Linfo_string4:
+	.asciz	"main"                          # string offset=128
+	.section	.debug_line,"",@progbits
+.Lline_table_start0:
+	.long	.Lunit_end - .Lunit_start     # unit length
+.Lunit_start:
+	.short	4   # version
+	.long	.Lprologue_end - .Lprologue_start # header length
+.Lprologue_start:
+	.byte	1                                      # minimum_instruction_length
+	.byte	1                                      # maximum_operations_per_instruction
+	.byte	0                                      # default_is_stmt
+	.byte	-5                                     # line_base
+	.byte	14                                     # line_range
+	.byte	13                                     # opcode_base
+	.byte	0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1     # arguments in standard opcodes
+	.byte	0                                      # end of include directories
+	.asciz	"main.c"                             # filename
+	.byte	0                                      # directory index
+	.byte	0                                      # modification time
+	.byte	0                                      # length of file (unavailable)
+	.byte	0                                      # end of filenames
+.Lprologue_end:
+	.byte	0x00, 9, 2                             # DW_LNE_set_address
+	.quad	0x1710                                 # Address Value
+	.byte	0x01                                   # DW_LNS_copy
+	.byte	0x49                                   # (address += 4,  line += -1,  op-index += 0)
```

Comment:
```suggestion
	.byte	0x49                                   # (address += 4,  line += -1,  op-index += 0)
        .byte 5                                        # DW_LNS_set_column
        .byte 2                                        # Column 2
```

Let's set the column here to distinguish the 0x1719 row from the 0x1710 row.

Reply:
Dosen't work. Produces this line table. 
```
Address            Line   Column File   ISA Discriminator OpIndex Flags
------------------ ------ ------ ------ --- ------------- ------- -------------
0x0000000000001710      1      0      1   0             0       0
0x0000000000001714      0      0      1   0             0       0
0x0000000000001719      1      2      1   0             0       0
0x000000000000171b      1      2      1   0             0       0  end_sequence
0x00000000000016c0      0      0      1   0             0       0
0x00000000000016cf      2      0      1   0             0       0
0x00000000000016d4      0      0      1   0             0       0
0x00000000000016d9      0      0      1   0             0       0
0x00000000000016df      0      0      1   0             0       0  end_sequence
```

Reply:
If the change is just for distinguishing. Why      adding here is necessary with a arbitrary column value?

Reply:
Maybe I'm missing something. My thoughts are that for the "WITHIN-SEQ" 0x1717 testcase, for the current table we would get the following output if --skip-line-zero had chosen the 0x1710 row..
```
foo
main.c:1:0 (approximate)
```
..and the same output if it had chosen the 0x1719 row.
With the proposed table we would get that same output if the 0x1710 row was chosen but..
```
foo
main.c:1:2 (approximate)
```
..if the 0x1719 row was chosen.

Reply:
> Maybe I'm missing something. My thoughts are that for the "WITHIN-SEQ" 0x1717 testcase, for the current table we would get the following output if --skip-line-zero had chosen the 0x1710 row..
> 
> ```
> foo
> main.c:1:0 (approximate)
> ```
>
> ..and the same output if it had chosen the 0x1719 row. With the proposed table we would get that same output if the 0x1710 row was chosen but..

That is not the case with 0x1719. In the current line-table which is :- 
```
Address            Line   Column File   ISA Discriminator OpIndex Flags
------------------ ------ ------ ------ --- ------------- ------- -------------
0x0000000000001710      1      0      1   0             0       0
0x0000000000001714      0      0      1   0             0       0
0x0000000000001719      1      0      1   0             0       0
0x000000000000171b      1      0      1   0             0       0  end_sequence
0x00000000000016c0      0      0      1   0             0       0
0x00000000000016cf      2      0      1   0             0       0
0x00000000000016d4      0      0      1   0             0       0
0x00000000000016d9      0      0      1   0             0       0
0x00000000000016df      0      0      1   0             0       0  end_sequence
```
```
Symbolizer o/p
~$ llvm-symbolizer --obj=skip-line-zero.o --skip-line-zero 0x1719
foo
main.c:1:0
```
This is because in the line table we have found the non-zero line-entry first , this will not be an approximate value. 
Check the lit test ```NO-APPROX```. 

> ```
> foo
> main.c:1:2 (approximate)
> ```
> 
> ..if the 0x1719 row was chosen.

The proposed line-table produced below having column entry as per your viewpoint.
```
Address            Line   Column File   ISA Discriminator OpIndex Flags
------------------ ------ ------ ------ --- ------------- ------- -------------
0x0000000000001710      1      0      1   0             0       0
0x0000000000001714      0      0      1   0             0       0
0x0000000000001719      1      2      1   0             0       0
0x000000000000171b      1      2      1   0             0       0  end_sequence
0x00000000000016c0      0      0      1   0             0       0
0x00000000000016cf      2      0      1   0             0       0
0x00000000000016d4      0      0      1   0             0       0
0x00000000000016d9      0      0      1   0             0       0
0x00000000000016df      0      0      1   0             0       0  end_sequence
```
```
Symbolizer output will be
~$ llvm-symbolizer --obj=skip-line-zero1.o 0x1719 --skip-line-zero
foo
main.c:1:2
```
The address 0x1719 have non-zero line value so it would not be an 'approximate'.

Reply:
I think @bd1976bris is referring to the test case above:
```
# RUN: llvm-symbolizer --obj=%t.o --skip-line-zero 0x1717 | FileCheck --strict-whitespace --match-full-lines --check-prefix=WITHIN-SEQ %s
```
Address 0x1717 is not an exact match for any line, so it would end up picking one of the line entries at 0x1710 or 0x1719. The output will look identical in either case, I believe, so the test could pass but be matching the wrong line. By changing the column value for 0x1719, it'll be obvious if the wrong line is matched.

Reply:
> Address 0x1717 is not an exact match for any line, so it would end up picking one of the line entries at 0x1710 or 0x1719. 

The logic of approximation is before only. So, it will search for line table entry in backwards uptill 0x1710.

Reply:
Yes, that's the theory of course, but if the code always did exactly what we wanted it to do, we'd not need test cases!

Reply:
Done.